app/services/katello/pulp3/repository.rb

```diff
@@ -91,16 +91,21 @@ def update_remote
       end
 
       def remote_partial_update
-        if remote_options[:url]&.start_with?('uln')
-          api.remotes_uln_api.partial_update(repo.remote_href, remote_options)
-        else
-          api.remotes_api.partial_update(repo.remote_href, remote_options)
+        url_type = remote_options[:url]&.start_with?('uln') ? 'uln' : 'default'
+        remote_type = repo.remote_href.start_with?('/pulp/api/v3/remotes/rpm/uln/') ? 'uln' : 'default'
+        href = repo.remote_href
+
+        if url_type == remote_type
+          api.get_remotes_api(href: href).partial_update(href, remote_options)
+        else # We need to recreate a remote of the correct type!
+          create_remote
+          delete_remote(href: href)
```

Comment:
FYI: This is what actually fixes the bug.

All the rest of the changes are essentially a refactor aimed at improving readability and robustness, using the correct source of truth for the remote type (in most cases the remote href itself), and also de-duplicating the remote API selection code.